### PR TITLE
Feature/qes o auth extentions

### PR DESCRIPTION
**Pull request by n0900**
Adds extensions as specified by CSC and "POTENTIAL consortium D3.1: UC Specification WP3" to allow handling of QES flows

### CODE REVIEW

File: openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt [hunk 1]
```diff
@@ -258,6 +266,82 @@ data class AuthenticationRequestParameters(
      */
     @SerialName("code_challenge_method")
     val codeChallengeMethod: String? = null,
+
+    //CSC additionally defines the following parameters
```
Comment (nodh) [inline openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt:270]: Imho this comment will never be read as it does not show up in the generated documentation for each property below.

File: openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt [hunk 2]
```diff
@@ -267,4 +351,23 @@ data class AuthenticationRequestParameters(
...
+ * CSC: Multiple hash values can be passed as comma separated values,
+ * e.g. oauth2/authorize?hash=dnN3ZX.. .ZmRm,ZjIxM3… Z2Zk,…
+ */
+object HashesSerializer : KSerializer<List<ByteArray>> {
+    override val descriptor: SerialDescriptor =
+        PrimitiveSerialDescriptor("HashesSerializer", PrimitiveKind.STRING)
+
+    override fun deserialize(decoder: Decoder): List<ByteArray> {
+        val listOfHashes = decoder.decodeString().split(",")
+        return listOfHashes.map { at.asitplus.dif.jsonSerializer.decodeFromString(ByteArrayBase64Serializer, it) }
```
Comment (nodh) [inline openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt:366]: but the documentation for `hashes` says it's base64-URL encoded

File: openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt [hunk 3]
```diff
@@ -267,4 +351,23 @@ data class AuthenticationRequestParameters(
             jsonSerializer.decodeFromString<AuthenticationRequestParameters>(it)
         }.wrap()
     }
+}
+
+/**
+ * CSC: Multiple hash values can be passed as comma separated values,
+ * e.g. oauth2/authorize?hash=dnN3ZX.. .ZmRm,ZjIxM3… Z2Zk,…
+ */
+object HashesSerializer : KSerializer<List<ByteArray>> {
```
Comment (nodh) [inline openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt:360]: Please move to separate file

File: vck-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidvci/WalletService.kt [hunk 4]
```diff
@@ -220,6 +216,40 @@ class WalletService(
...
+                codeChallenge = generateCodeVerifier(state),
+                codeChallengeMethod = CODE_CHALLENGE_METHOD_SHA256,
+            )
+
+            is AuthorizationDetails.CSCCredential -> AuthenticationRequestParameters(
+                responseType = GRANT_TYPE_CODE,
+                state = state,
+                clientId = clientId,
+                authorizationDetails = setOf(authorizationDetails),
+                scope = "credential",
```
Comment (nodh) [inline vck-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidvci/WalletService.kt:245]: seems like "credential" should be a constant

File: vck-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidvci/WalletService.kt [hunk 5]
```diff
@@ -300,6 +330,34 @@ class WalletService(
         )
     }
 
+    /**
+     * CSC: Adds the possibility for CSC requests. TODO should be list of AuthorizationDetails?
```
Comment (nodh) [inline vck-openid/src/commonMain/kotlin/at/asitplus/wallet/lib/oidvci/WalletService.kt:334]: So is it a set of authorization details or not? :-)

**Review by nodh — CHANGES_REQUESTED** [review]
Nice, only minor remarks

File: openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt (hunk 2 above)
Comment (n0900) [reply openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt]: According to CSC-API v2.0.0.2 hashes are base64 encoded. Just double checked

Comment (n0900) [reply openid-data-classes/src/commonMain/kotlin/at/asitplus/openid/AuthenticationRequestParameters.kt]: If this is inconsistent with the RQes spec then I think I prefer the CSC version as it is likely to change to CSC soon

**Review by nodh — APPROVED** [review]

**Pull request merged**